sound/soc/sof/ipc.c

```diff
@@ -416,6 +416,19 @@ static void ipc_trace_message(struct snd_sof_dev *sdev, u32 msg_id)
 	}
 }
 
+/*
```

Comment:
comment on commit message:

Position update from dsp might cause alsa calling stop and sof
eventually sending ipc to dsp. This can happen if the stream is in
draining or xrun state.

I think you omitted something here. isn't this the application that calls stop? Or is this an xrun handled at the ALSA level? Also are you trying to say a first position update causes a state change, but there is still a period elapsed received after the  state change? or that there is a race condition with the state change happening in the middle of the position update, without the IPC completing with a reply. In the latter case, we should first complete the IPC at the Intel level then notify higher levels.

Reply:
aplay with ctrl-c will cause drain to be called. this will move the stream to draining state. it would call TRIGGER_DRAIN but we don't handle that, it is not even setup in ops. As I understand it alsa core's drain will wait in a loop for position update to end draining. Now our position update comes, in alsa core it notices that state is draining -> will call stop (we are in irq handling with busy bit on) -> ack to stop is never received

Reply:
so position update in alsa core will check for draining and xrun states and just call stop (in same context, no deferral processing) and then we are in trouble

Reply:
https://github.com/thesofproject/linux/issues/66 exists since August 7, what can I say...

Reply:
@plbossart hmm yep 66 should have been taken care of earlier. However, I think we would still have to make sure that we would not allow irq handler eventually call ipc towards dsp in any circumstances? As I understand the alsa logic is to wait for position update to "finish" drain. if the position update is not doing that there's like 10s timeout, which will end up calling stop and different timeout error in dmesg.

I'm just saying that handling TRIGGER_DRAIN would not solve this issue totally... We would still have a race with position update coming in irq handler from dsp.  And what should we do with TRIGGER_DRAIN? we could set some state in driver and try to filter out certain messages (like stop) in this situation. I tried this and it fixes the situation, because even if you filter out stop in draining state, alsa will move the stream state to stop and eventually second stop will come from aplay. However, I I think the logic is then wrong from what alsa expects, because dsp is running happily between 2 stops.

Anyway, I could try to bury this to intel layer, as I see this is intel/byt specific. I don't know how our other platforms are handling position updates. I have to then refactor little bit the general ipc.c file

Reply:
@plbossart, I tend to agree with @juimonen here. Further, I'm not sure this is Intel specific. The 
loop in generic ipc code where we:

```
irq-thread -> ipc.c:snd_sof_ipc_msgs_rx() 
     -> ipc.c:ipc_stream_message() 
     -> pcm_lib.c:snd_pcm_period_elapsed()  /* all normal so far */
     -> snd_pcm_do_stop() /* this can happen e.g. when drain is complete */ 
     -> pcm.c:sof_pcm_trigger() 
     -> ipc.c:sof_ipc_tx_message() -> ... -> ipc.c:tx_wait_done() 
     -> wait_event_timeout()  /* yikes, we are sleeping in IRQ handler thread! */
'''

@lyakh is out this week, so not sure whether this is on his todo for IPC, but this seems a problem in the general IPC code and you just happen to hit this on BYT.

An alternative fix would be to modify tx_wait() so that it behaves differently when called within the IRQ handler like in above calltrace (and not wait in this case).

Reply:
My comment was on the 'with ipc busy bit enabled' comment. Maybe it's a generic problem, but I want to make sure we keep the Intel parts contained to the intel/ folder.

Reply:
@plbossart @kv2019i I see @keyonjie has done exactly same thing as I have done here in his own topic/sof-dev branch for fixing ipc issues in other platforms: https://github.com/keyonjie/linux/tree/topic/sof-dev. 

So this seems to be not only byt specific. @keyonjie initialization of "work" is much more cleaner than mine, but he's calling deferred work every time, me only for certain pcm state. I can consolidate with @keyonjie if we agree that we can call deferred work all the time (so extend to all platforms).

@kv2019i I tried to catch the "irqness" in tx_wait with in_interrupt(), but that didn't work, probably because we are in the "thread" part of threaded irq. So for me it seems that @keyonjie work is the most concise (if work queue usage is allowed for every elapsed call). It will for sure guarantee there will be no send in irq context.

Reply:
> @plbossart, I tend to agree with @juimonen here. Further, I'm not sure this is Intel specific. The
> loop in generic ipc code where we:
> 
> ```
> irq-thread -> ipc.c:snd_sof_ipc_msgs_rx() 
>      -> ipc.c:ipc_stream_message() 
>      -> pcm_lib.c:snd_pcm_period_elapsed()  /* all normal so far */
>      -> snd_pcm_do_stop() /* this can happen e.g. when drain is complete */ 
>      -> pcm.c:sof_pcm_trigger() 
>      -> ipc.c:sof_ipc_tx_message() -> ... -> ipc.c:tx_wait_done() 
>      -> wait_event_timeout()  /* yikes, we are sleeping in IRQ handler thread! */
> '''
> 
> @lyakh is out this week, so not sure whether this is on his todo for IPC, but this seems a problem in the general IPC code and you just happen to hit this on BYT.
> 
> An alternative fix would be to modify tx_wait() so that it behaves differently when called within the IRQ handler like in above calltrace (and not wait in this case).
> ```

@kv2019i yes that is ideal, but as I have explained in the mail thread, sleep/wait until the TX is acknowledged from other processor is mandatory in our current design, changing that is a big surgery, we should do that in IPC 2.0

Maybe @lgirdwood can chime into this.